libvast/src/detail/sigma.cpp

```diff
@@ -0,0 +1,369 @@
+      dis.emplace_back(std::move(con[0]));
+    else
+      dis.emplace_back(std::move(con));
+    return dis.size() == 1 ? std::move(dis[0]) : expression{dis};
+  };
+
+  template <class Iterator>
+  bool parse(Iterator& f, const Iterator& l, expression& result) const {
+    using namespace parser_literals;
+    auto ws = ignore(*parsers::space);
+    auto negate = [](expression x) { return negation{std::move(x)}; };
+    rule<Iterator, expression> expr;
+    rule<Iterator, expression> group;
+    // clang-format off
+    group
+      = '(' >> ws >> ref(expr) >> ws >> ')'
+      | "not"_p >> ws >> '(' >> ws >> (ref(expr) ->* negate) >> ws >> ')'
+      | "not"_p >> ws >> search_id ->* negate
+      | search_id
+      ;
+    auto and_or
+      = "or"_p  ->* [] { return bool_operator::logical_or; }
+      | "and"_p  ->* [] { return bool_operator::logical_and; }
+      ;
+    expr
+      = (group >> *(ws >> and_or >> ws >> ref(group)) >> ws) ->* to_expr
+      ;
+    // clang-format on
+    auto p = expr >> parsers::eoi;
+    return p(f, l, result);
+  }
+
+  search_id_symbol_table search_id;
+};
+
+// The following invariants apply according to the Sigma spec:
+// - All values are treated as case-insensitive strings
+// - You can use wildcard characters '*' and '?' in strings
+// - Wildcards can be escaped with \, e.g. \*. If some wildcard after a
+//   backslash should be searched, the backslash has to be escaped: \\*.
+// - Regular expressions are case-sensitive by default
+// - You don't have to escape characters except the string quotation
+//   marks '
+caf::optional<pattern> make_pattern(std::string_view str) {
+  auto f = str.begin();
+  auto l = str.end();
+  std::string rx;
+  // FIXME: this is a pretty hand-wavy approach to transforming a glob string
+  // to a valid regex. We need to revisit this once we have actual pattern
+  // support in the query language.
+  while (f != l) {
+    auto c = *f++;
+    if (c == '\\') {
+      if (f == l) {
+        rx += '\\'; // A single backslash at the end is a literal backslash.
+        break;
+      }
+      auto w = *f++;
+      if (w == '*') {
+        // Escaped wildcard
+        rx += '*';
+      } else if (w == '?') {
+        // Escaped wildcard
+        rx += '?';
+      } else if (w == '\\') {
+        // Double backslash (\\). We include at least one backslash.
+        rx += '\\';
+        // If no wildcard follows \\, we have a literal double backslash. If a
+        // wildcard follows, e.g., \\* or \\?, then we have a single
+        // backslash plus wildcard.
+        if (f == l)
+          rx += '\\';
+      } else {
+        // Do nothing by default;
+        rx += '\\';
+        rx += w;
+      }
+    } else if (c == '*') {
+      rx += ".*";
+    } else if (c == '?') {
+      rx += ".";
+    } else {
+      rx += c;
+    }
+  };
+  // It's only a pattern if it differs from a regular string.
+  if (str == rx)
+    return caf::none;
+  return pattern{std::move(rx)};
```

Comment:
I think you need to prepend the pattern with `^` and append it with `$`.

Reply:
Since patterns are not yet support on field values, this is more or less irrelevant. We need to revisit the pattern generation anyway *after* we can do pattern matching.

Reply:
Please add a comment then.